### PR TITLE
#34 Adding annotation processor for exposing referenced stream names;

### LINKED ISSUE #34: Provide means to expose stream names accessed by a custom pipeline

### PR DESCRIPTION
**Pull request by gunnarmorling**
The file META-INF/decodable/stream-names.properties will be generated based on the InputStreams and OutputStreams annotations.

Fixes #34

### CODE REVIEW

File: examples/apache-maven/custom-pipelines-hello-world/src/main/java/co/decodable/examples/cpdemo/DataStreamJob.java [hunk 1]
```diff
@@ -19,21 +22,28 @@
 
 import co.decodable.sdk.pipeline.DecodableStreamSink;
 import co.decodable.sdk.pipeline.DecodableStreamSource;
+import co.decodable.sdk.pipeline.metadata.SinkStreams;
+import co.decodable.sdk.pipeline.metadata.SourceStreams;
 
+@SourceStreams(PURCHASE_ORDERS_STREAM)
+@SinkStreams(PURCHASE_ORDERS_PROCESSED_STREAM)
```
Comment (rmetzger) [inline examples/apache-maven/custom-pipelines-hello-world/src/main/java/co/decodable/examples/cpdemo/DataStreamJob.java:29]: That's a nice solution!

The downside is of course that you have to define the stream names at compile time .. but anything else is orders of magnitudes more complex to implement in the backend.

File: sdk/src/test/java/co/decodable/sdk/pipeline/internal/config/metadata/MetadataProcessorTest.java [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+
+  private static URL getJobSourceFile() {
+    try {
+      return new File(
+              "./src/test/java/co/decodable/sdk/pipeline/snippets/PurchaseOrderProcessingJob.java")
+          .toURI()
+          .toURL();
+    } catch (IOException e) {
+      throw new RuntimeException(e);
+    }
```
Comment (rmetzger) [inline sdk/src/test/java/co/decodable/sdk/pipeline/internal/config/metadata/MetadataProcessorTest.java:49]: What speaks against declaring the IOException at the test method?

**Review by rmetzger — APPROVED** [review]
Nice change.

What's a good place to document the file format and the annotation mechanism? If folks want to generate the metadata file themselves (e.g. they don't want to specify the stream names at compile time), ...

File: examples/apache-maven/custom-pipelines-hello-world/src/main/java/co/decodable/examples/cpdemo/DataStreamJob.java (hunk 1 above)
Comment (gunnarmorling) [reply examples/apache-maven/custom-pipelines-hello-world/src/main/java/co/decodable/examples/cpdemo/DataStreamJob.java:29]: Not necessarily compile time, rather assembly time. The contract is the _META-INF/decodable/stream-names.properties_ file, not the annotations. In addition, we could allow overriding at deployment or activation time, but I'd cross that bridge when there's actual demand for this raised by users.

Ultimately, I think the SDK should report stream names dynamically at runtime, but again we can improve this story iteratively.

File: sdk/src/test/java/co/decodable/sdk/pipeline/internal/config/metadata/MetadataProcessorTest.java (hunk 2 above)
Comment (gunnarmorling) [reply sdk/src/test/java/co/decodable/sdk/pipeline/internal/config/metadata/MetadataProcessorTest.java]: Reworked.

**Comment by gunnarmorling** [comment]
> What's a good place to document the file format and the annotation mechanism? If folks want to generate the metadata file themselves (e.g. they don't want to specify the stream names at compile time), ...

Yeah, I've been wondering about the same, too. For now, I've added this to the JavaDoc:

Screenshot 2023-07-18 at 15-34-27 co decodable sdk pipeline (Decodable Pipeline SDK 1 0 0-SNAPSHOT)

Ultimately, we should provide a reference guide for the SDK (CC @matsukawa-maria), but I think for now it's good enough to have everything comprehensively documented in the API docs themselves.

**Review by rmetzger — APPROVED** [review]
Thx for addressing my comments

**Pull request merged**